File: examples/school/go/adapters.go
package main

import (
	"encoding/json"
	"net/http"
	"strings"

	"github.com/gin-gonic/gin"
	"github.com/go-chi/chi/v5"
)

// StdlibServer Serves the Routes with net/http alone.
// Reference: https://pkg.go.dev/net/http#ServeMux
type StdlibServer struct{}

// ServeRoutes Mounts every Route on a stdlib Mux.
func (StdlibServer) ServeRoutes(routes []Route, address string) error {
	mux := http.NewServeMux()

	for _, route := range routes {
		mux.HandleFunc(route.Method+" "+route.Pattern, StdlibHandlerFor(route))
	}

	return http.ListenAndServe(address, mux)
}

// StdlibHandlerFor Wraps one Route into a stdlib HandlerFunc.
func StdlibHandlerFor(route Route) http.HandlerFunc {
	names := ListPatternParams(route.Pattern)

	return func(w http.ResponseWriter, r *http.Request) {
		path := map[string]string{}
		for _, name := range names {
			path[name] = r.PathValue(name)
		}

		WriteReplyAsJSON(w, route.Handle(ReadRequestValues(r, path)))
	}
}

// ChiServer Serves the same Routes with chi.
// Reference: https://pkg.go.dev/github.com/go-chi/chi/v5
type ChiServer struct{}

// ServeRoutes Mounts every Route on a chi Router.
func (ChiServer) ServeRoutes(routes []Route, address string) error {
	router := chi.NewRouter()

	for _, route := range routes {
		router.MethodFunc(route.Method, route.Pattern, ChiHandlerFor(route))
	}

	return http.ListenAndServe(address, router)
}

// ChiHandlerFor Wraps one Route, Reading Params the chi Way.
func ChiHandlerFor(route Route) http.HandlerFunc {
	names := ListPatternParams(route.Pattern)

	return func(w http.ResponseWriter, r *http.Request) {
		path := map[string]string{}
		for _, name := range names {
			path[name] = chi.URLParam(r, name)
		}

		WriteReplyAsJSON(w, route.Handle(ReadRequestValues(r, path)))
	}
}

// GinServer Serves the same Routes with gin.
// Gin Owns its own Context, so this Adapter Carries the most Work.
// Reference: https://pkg.go.dev/github.com/gin-gonic/gin
type GinServer struct{}

// ServeRoutes Mounts every Route on a gin Engine.
func (GinServer) ServeRoutes(routes []Route, address string) error {
	gin.SetMode(gin.ReleaseMode)
	engine := gin.New()

	for _, route := range routes {
		engine.Handle(route.Method, TranslateRoutePattern(route.Pattern), GinHandlerFor(route))
	}

	return engine.Run(address)
}

// GinHandlerFor Wraps one Route, Reading Params the gin Way.
func GinHandlerFor(route Route) gin.HandlerFunc {
	names := ListPatternParams(route.Pattern)

	return func(c *gin.Context) {
		path := map[string]string{}
		for _, name := range names {
			path[name] = c.Param(name)
		}

		WriteReplyAsJSON(c.Writer, route.Handle(ReadRequestValues(c.Request, path)))
	}
}

// TranslateRoutePattern Turns {id} into :id, which is all gin Wants.
func TranslateRoutePattern(pattern string) string {
	parts := strings.Split(pattern, "/")

	for index, part := range parts {
		if strings.HasPrefix(part, "{") && strings.HasSuffix(part, "}") {
			parts[index] = ":" + strings.Trim(part, "{}")
		}
	}

	return strings.Join(parts, "/")
}

// WriteReplyAsJSON is the one Place that Touches a ResponseWriter.
func WriteReplyAsJSON(w http.ResponseWriter, reply Response) {
	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(reply.Status)

	if reply.Body != nil {
		_ = json.NewEncoder(w).Encode(reply.Body)
	}
}

// SelectServerAdapter Picks the Framework at Startup, never at Compile Time.
func SelectServerAdapter(name string) Server {
	switch name {
	case "gin":
		return GinServer{}
	case "chi":
		return ChiServer{}
	default:
		return StdlibServer{}
	}
}

File: examples/school/ts/src/adapters.ts
import { createServer, type IncomingMessage, type ServerResponse } from "node:http";

import express from "express";
import Fastify from "fastify";

import { listPatternParams, readBearerToken, type Request, type Response, type Route, type Server } from "./transport.js";

// NodeServer Serves the Routes with node:http alone.
// Reference: https://nodejs.org/api/http.html#httpcreateserveroptions-requestlistener
export class NodeServer implements Server {
  async serveRoutes(routes: Route[], port: number): Promise<void> {
    const server = createServer((req, res) => this.dispatchNodeRequest(routes, req, res));

    await new Promise<void>((resolve) => server.listen(port, resolve));
  }

  // dispatchNodeRequest Matches by hand, because node:http Has no Router.
  private async dispatchNodeRequest(routes: Route[], req: IncomingMessage, res: ServerResponse) {
    const url = new URL(req.url ?? "/", "http://localhost");
    const found = matchRoutePattern(routes, req.method ?? "GET", url.pathname);

    if (!found) return writeReplyAsJSON(res, { status: 404, body: { error: "no such Route" } });

    const body = await readRequestStream(req);
    writeReplyAsJSON(res, found.route.handle({
      path: found.path,
      query: Object.fromEntries(url.searchParams),
      token: readBearerToken(req.headers.authorization),
      body,
    }));
  }
}

// ExpressServer Serves the same Routes with express.
// Reference: https://expressjs.com/en/5x/api.html
export class ExpressServer implements Server {
  async serveRoutes(routes: Route[], port: number): Promise<void> {
    const app = express();
    app.use(express.text({ type: "*/*" }));

    for (const route of routes) {
      app[route.method.toLowerCase() as "get"](translateRoutePattern(route.pattern), (req, res) => {
        const reply = route.handle({
          path: req.params as Record<string, string>,
          query: req.query as Record<string, string>,
          token: readBearerToken(req.headers.authorization),
          body: typeof req.body === "string" ? req.body : "",
        });

        res.status(reply.status).json(reply.body ?? undefined);
      });
    }

    await new Promise<void>((resolve) => { app.listen(port, () => resolve()); });
  }
}

// FastifyServer Serves the same Routes with fastify.
// Reference: https://fastify.dev/docs/latest/Reference/Routes/
export class FastifyServer implements Server {
  async serveRoutes(routes: Route[], port: number): Promise<void> {
    const app = Fastify();

    // Fastify Parses JSON before we See it, so its Parser Comes out first.
    // Reference: https://fastify.dev/docs/latest/Reference/ContentTypeParser/
    app.removeAllContentTypeParsers();
    app.addContentTypeParser("*", { parseAs: "string" }, (_, body, done) => done(null, body));

    for (const route of routes) {
      app.route({
        method: route.method,
        url: translateRoutePattern(route.pattern),
        handler: async (req, reply) => {
          const answer = route.handle({
            path: req.params as Record<string, string>,
            query: req.query as Record<string, string>,
            token: readBearerToken(req.headers.authorization),
            body: typeof req.body === "string" ? req.body : "",
          });

          return reply.code(answer.status).send(answer.body ?? undefined);
        },
      });
    }

    await app.listen({ port });
  }
}

// translateRoutePattern Turns {id} into :id, which express and fastify Want.
export function translateRoutePattern(pattern: string): string {
  return pattern
    .split("/")
    .map((part) => (part.startsWith("{") && part.endsWith("}") ? `:${part.slice(1, -1)}` : part))
    .join("/");
}

// matchRoutePattern is the Router node:http Does not Ship.
export function matchRoutePattern(routes: Route[], method: string, pathname: string) {
  for (const route of routes) {
    const names = listPatternParams(route.pattern);
    const shape = new RegExp(`^${route.pattern.replace(/\{[^}]+\}/g, "([^/]+)")}$`);
    const found = route.method === method ? shape.exec(pathname) : null;

    if (found) {
      const path = Object.fromEntries(names.map((name, index) => [name, found[index + 1]!]));
      return { route, path };
    }
  }

  return null;
}

// readRequestStream Collects the Body node:http Delivers in Pieces.
function readRequestStream(req: IncomingMessage): Promise<string> {
  return new Promise((resolve) => {
    let text = "";
    req.on("data", (chunk) => { text += chunk; });
    req.on("end", () => resolve(text));
  });
}

// writeReplyAsJSON is the one Place that Touches a ServerResponse.
function writeReplyAsJSON(res: ServerResponse, reply: Response) {
  res.writeHead(reply.status, { "Content-Type": "application/json" });
  res.end(reply.body === null ? "" : JSON.stringify(reply.body));
}

// selectServerAdapter Picks the Framework at Startup, never at Build Time.
export function selectServerAdapter(name: string): Server {
  if (name === "express") return new ExpressServer();
  if (name === "fastify") return new FastifyServer();

  return new NodeServer();
}
